src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java

```diff
@@ -502,23 +573,32 @@ public void createDirectory(Path dir,
         }
 
         // create bucket if necessary
-        Bucket bucket = s3Path.getFileStore().getBucket();
-        String bucketName = s3Path.getFileStore().name();
+        final Bucket bucket = s3Path.getFileStore().getBucket();
+        final String bucketName = s3Path.getFileStore().name();
+
         if (bucket == null)
         {
-            s3Path.getFileSystem().getClient().createBucket(bucketName);
+            //TODO: How to get s3fs_region property value?
+            final Region region = Region.US_EAST_1;
+            final CreateBucketConfiguration configuration = CreateBucketConfiguration.builder()
+                                                                                     .locationConstraint(region.id())
+                                                                                     .build();
+            final CreateBucketRequest request = CreateBucketRequest.builder()
+                                                                   .bucket(bucketName)
+                                                                   .createBucketConfiguration(configuration)
+                                                                   .build();
+            client.createBucket(request);
         }
 
         // create the object as directory
-        ObjectMetadata metadata = new ObjectMetadata();
-        metadata.setContentLength(0);
+        final String directoryKey = s3Path.getKey().endsWith("/") ? s3Path.getKey() : s3Path.getKey() + "/";
```

Comment:
It is possible to create valid S3 paths like `s3://mybucket/folder/./folder2/../folder3/object1` which would have a different interpretation in POSIX filesystems. FSx for Lustre treats paths like POSIX paths so the above path would be effectively invisible which is probably the right thing to do even though it means that there will be some incompatibility between S3 and nio for esoteric edge cases.

Reply:
@markjschreiber, so what should be the right solution here?

Reply:
There isn't a 100% compatible way to do it. I would go with how FSx for Lustre seems to do it. I think most people would expect s3fs to behave in a POSIX fashion and cases where people create S3 objects with paths/keys like the one above would be rare.